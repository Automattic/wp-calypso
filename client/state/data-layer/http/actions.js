/**
 * Internal dependencies
 */
import { HTTP_REQUEST } from 'state/action-types';

/***
 * @typedef {Object} RequestDescription
 * @property {String} url the url to request
 * @property {String} method the method we should use in the request: GET, POST etc.
 * @property {Array<Array<String>>} headers array of [ 'key', 'value' ] pairs for the request headers
 * @property {Object} body data send as body
 * @property {Boolean} withCredentials save cookie set on request
 * @property {Object|Function} onSuccess action to dispatch on success with data meta
 * @property {Object|Function} onFailure action to dispatch on failure with error meta
 * @property {Object} options other options
 */

/***
 * Creates a raw http action request
 *
 * @param {RequestDescription} request HTTP request description
 * @param {?Object} action default action to call on HTTP events
 * @returns {Object} raw http action
 */
export const rawHttp = ( {
	url,
	method,
	headers,
	body,
	withCredentials,
	onSuccess,
	onFailure,
	...options,
}, action = null ) => ( {
	type: HTTP_REQUEST,
	url,
	method,
	headers,
	body,
	withCredentials,
	onSuccess: onSuccess || action,
	onFailure: onFailure || action,
	options,
} );
