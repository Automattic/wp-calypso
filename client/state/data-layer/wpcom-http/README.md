# WordPress.com API HTTP Layer

## Summary

WordPress.com API requests require an initial translation from understanding what data pieces of Calypso might be wanting and the endpoints on WordPress.com which can provide that data.
Once that translation has been made the requests have to be sent out over an HTTP connection.
This middleware layer provides a foundation under the API data layer which can be used to pass around pure and stateless descriptions of those desired HTTP requests.
By removing the burden of directly issuing and responding to HTTP requests in the API layer and moving it here, we can more easily accommodate global and centralized policies on those requests, such as batching, dropping, and retrying on failure.

In contrast to a fully-generic HTTP layer this one understands that requests are being made directly to the WordPress.com API and allowances are made in the HTTP Request descriptions to acknowledge this and make working with it a smoother process; for example, authentication is automatically handled.

## Why?

The end-goal for this two-layer middleware approach is to make offline-use first-class in Calypso and to enable things like mobile data management and battery optimization.
However, it also brings about many secondary benefits to the way we build the data gathering and updating bits inside of Calypso.

Let's look at how this system works from a high-level perspective.

## Overview

<!-- the following diagram was generated in draw.io - it can be edited by pasting in the contents of the SVG itself -->
![](https://cldup.com/unQOzvDkjtq/BpewKu.svg)

Wow, that can look confusing!
While this isn't promising to make everything easier, it _is_ promising to make things worth while.
And in fact, with a little bit of exposure it reveals a pattern that ends up simpler and will probably seem easier than "the old way."

### Step A: Component needs to update data

A React component wants to indicate that a given post ought to be marked as liked.
Its API includes two functions for accomplishing this: `likePost( siteId, postId )` and `unlikePost( siteId, postId )`.
It knows nothing about how those two functions work **and it doesn't need to know anything about how they work**, only that they are the names it uses to accomplish its goal.
Because the component interacts with the data layer at this data-centric interface we have the ability to change the implementations for how we interact with synchronizing our data without demanding that the components themselves change.
It calls `likePost` after the user clicks on the like button and that function dispatches a Redux action carrying that intention.

### Step B: API middleware intercepts action

The WordPress.com API middleware understands what it means to like a post.
It knows that when we do so the server needs to be notified so that the like actually persists across browser sessions and API calls.
Further, it knows that it's possible that the server might reject the like, or some other trigger might cause the like to fail.

The middleware speaks with the WordPress.com servers through HTTP requests.
These requests have special properties: they can fail, we might want them to automatically retry upon failure, they can be batched, etc.
Therefore, instead of coding all of this logic into the API middleware itself the middleware will issue another declarative action describing the HTTP request and the meta information and policies about that request.
The new action carries along the information for the request such as the WordPress.com API path and associated query data, an action to dispatch if the request is a success, and an action to dispatch if the request fails. It may also carry an action to call when uploading a file in order to track progress updates through the process.

Because this middleware is merely attempting to keep the server in sync with Calypso's updated state, it needs to pass the action along so that the reducers get the chance to update the app state and reflect the change.

### Step C: HTTP middleware runs actual HTTP request

The HTTP middleware understands what it means to run HTTP requests.
It may take some liberties in how it decides to actually execute the requests; **this liberty is what allows us to provide the global policies like retry behavior, offline behavior, and batching**.

However the middleware decides to do it, it issues the HTTP request.
When either a response comes back in or the request fails, the middleware will dispatch a new follow-up action.
This responder action comes from the original one which triggered the network call (the `HTTP_REQUEST` action) and it will be wrapped with meta information including the response data or error message.
That new action gets re-injected into Redux and will likely be handled by more middleware before it hits the reducers.

In this case, we have assumed that a successful response from the server (a `2xx` status code) means that the like was successful.
This is a wrong assumption, but it is good enough for this discussion.
As seen in the diagram, the API middleware instructed the HTTP middleware to fire a `LIKE_POST` action if the HTTP request succeeds and an `UNLIKE_POST` action if it fails.
This `UNLIKE_POST` action will end up rolling back the first like, which in effect could not be completed because the server failed to update.

### Step D: Reducers see action and update app state

The actual `LIKE_POST` action (and possible the `UNLIKE_POST` as well) hits the reducers and the app state is updated.
This will trigger a re-render on the post component which started the chain and we come full circle.

## Implementation

What essentially is an HTTP request?
What makes it up?
How should it behave?
This implementation provides answers to these questions that we can use to standardize how we make them and deal with them.

> HTTP requests are indications of how we want what's happening in Calypso to change the external world or how we want the external world to change what's happening in Calypso.

They are asynchronous operations which can succeed and which can and do often fail.
They may return with data or error messages from server but they might also return without any data.
They may send data to a server and if it's a lengthy process - an upload - they may issue progress events to indicate how far into the process they currently are.
They are minimally defined by two properties - the HTTP method type (`GET` or `POST`) and the URL path where they should be sent.

Inside of this model we can build a simple interface for issuing and responding to HTTP requests.

### Usage

Use the `http` function to create a Redux action describing an HTTP request.

```js
import { http } from 'state/data-layer/wpcom-http/actions';

// announce presence - requires no response inside of Calypso
dispatch( http( {
	method: 'POST',
	path: '/me/fairly-land/announce-presence'
} ) );
```

This function accepts an array of parameters for further specifying information about the request.
Please see `state/data-layer/wpcom-http/actions.js` for more information.

Specify one or more of the following properties in an HTTP request in order to respond to the different lifecycle events of the request.

 - `onSuccess` dispatched when the request returns with a `2xx` status code
 - `onFailure` dispatched when the request fails for any reason whether for an error status code, network timeout, or otherwise
 - `onProgress` _if_ issuing a `POST` request _and_ uploading a file then this is dispatched on progress events

Please note that _these are not callback functions_ but rather normal Redux actions which can be serialized and deserialized, inspected and transformed, and monitored or logged.

```js
// get fictional blog updates
// we will ignore failures if they exist
dispatch( http( {
	method: 'GET',
	path: `/sites/${ siteId }/fairy-land/updates`,
	onSuccess: { type: ADD_UPDATE, siteId }
} ) );
```

You may need to know about the original action which triggered the request when the response comes back in.
Notice how we can store that information inside of the responder actions just like how we encapsulate date in closures.

```js
const missileMiddleware = ( store, action ) => {
	if ( FIRE_ZE_MISSILES !== action.type ) {
		return;
	}

	dispatch( http( {
		method: 'POST',
		path: '/foe/missile/new',
		onError: { type: TAKE_A_NAP, duration: 5 * MINUTES_IN_SECONDS, nextAction: FIRE_ZE_MISSILES }
	} ) );
};
```

We can be as creative or dull as we want to be in the request lifecycle.
Because the information from the request response extends the action through meta, we can actually re-use the originating action and handle it based on that meta.

```js
import { getProgress } from 'state/data-layer/wpcom-http/utils';

const packageMiddleware = ( store, action ) => {
	if ( CREATE_PACKAGE !== action.type ) {
		return;
	}

	const progress = getProgress( action );
	if ( progress ) {
		return dispatch( {
			type: UPDATE_PACKAGE_UPLOAD,
			packageId: action.packageId,
			progress
		} );
	}

	// reuse the action so we don't need to create
	// additional action types for _PROGRESS etc...
	return dispatch( http( {
		method: 'POST',
		path: '/packages/new',
		body: action.FileData,
		onProgress: action
	} ) );
}
```

If we decide to reuse the actions it can get tedious to write out the same action every time.
Therefore you can skip the responder actions and pass along the original action as the second and optional parameter to `http()`.

```js
dispatch( http( {
	method: 'GET',
	path: '/me/splines',
	query: {
		fields: 'ID,tremie_pipe_type'
	}
}, action ) );
```

If given, the action passed as the second and optional parameter will take the place of all unspecified `onSuccess`, `onFailure`, and `onProgress` responder events.
Because we have a very common pattern when issuing requests there is a built-in helper to hand each action off to the right function based on the (possibly) attached metadata.

```js
import { dispatchRequest } from 'state/data-layer/wpcom-http/utils';

// create request when no meta present, add on success, alert on failure
dispatchRequest( fetchMenu, addMenuItems, alertFailure );

// create request when no meta present, indicate on success, undo on failure, update on progress
dispatchRequest( sendRecipe, indicateSuccess, removeRecipe, updateRecipeProgress );
```

### Helpers

Most likely you will want to use the provided helper methods and structure your code into three pieces: a function which generates the descriptive HTTP requests; a function which handles successful responses; and a function which handles failing responses.
If you are uploading a file you will have a fourth function which handles progress events.

Each of these functions will take the normal middleware arguments; additionally the success, failure, and progress functions take an additional argument which is the response data or error respectively.
Please note that for this example we have included a progress event even though one would not normally exist for liking posts.
Its inclusion her is meant merely to illustrate how the pieces can fit together.

```js
// API Middleware, Post Like
import { LIKE_POST, LIKE_POST_PROGRESS, UNLIKE_POST } from 'state/action-types';
import { http } from 'state/data-layer/wpcom-http/actions';
import { QUEUE_REQUEST } from 'state/data-layer/wpcom-http/constants';

/**
 * @see https://developer.wordpress.com/docs/api/1.1/post/sites/%24site/posts/%24post_ID/likes/new/ API description
 */
const likePost = ( { dispatch }, action ) => {
	// dispatch intent to issue HTTP request
	// by not supplying onSuccess, onError, and onProgress
	// _and_ by supplying the second optional `action`
	// argument we will be reusing the original action to
	// follow-up with these requests, but when it comes
	// back it will be wrapped with meta information
	// describing the response from the HTTP events
	dispatch( http( {
		method: 'POST',
		path: `/sites/{ action.siteId }/posts/${ action.postId }/likes/new`,
		
		// indicate what should happen if we have
		// no network connection: queue to replay
		// when the network reconnects
		// (not implemented yet)
		whenOffline: QUEUE_REQUEST,
	}, action ) );
}

/**
 * Called on success from HTTP middleware with `data` parameter
 *
 * This is the place to map fromAPI to Calypso formats
 */
const verifyLike = ( { dispatch }, { siteId, postId } data ) => {
	// this is a response to data coming in from the data layer,
	// so skip further data-layer middleware with local
	dispatch( local( {
		type: data.i_like ? LIKE_POST : UNLIKE_POST,
		siteId,
		postId,
		likeCount: data.like_count,
	} ) );
}

/**
 * Called on failure from the HTTP middleware with `error` parameter
 */
const undoLike = ( { dispatch }, { siteId, postId }, error ) => {
	// skip data-layer middleware
	dispatch( local( {
		type: UNLIKE_POST,
		siteId,
		postId,
	} ) );
}

/**
 * Maps progress information from the API into a Calypso-native representation
 */
const updateProgress = ( store, { siteId, postId }, progress ) => {
	dispatch( {
		type: LIKE_POST_PROGRESS,
		siteId,
		postId,
		percentage: 100 * progress.loaded / ( progress.total + Number.EPSILON )
	} );
}

export default {
	// watch for this action       -> initiate, onSuccess,  onFailure, onProgress
	[ LIKE_POST ]: [ dispatchRequest( likePost, verifyLike, undoLike, updateProgress ) ],
};
```

It's important that perform the mapping stage when handling API request responses to go _from_ the API's native data format _to_ Calypso's native data format.
These middleware functions are the perfect place to perform this mapping because it will leave API-specific code separated into specific places that are relatively easy to find.
