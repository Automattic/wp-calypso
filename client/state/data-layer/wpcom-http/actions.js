/**
 * Internal dependencies
 */
import { WPCOM_HTTP_REQUEST } from 'state/action-types';

/**
 * Returns a valid WordPress.com API HTTP Request action object
 *
 * @param {string} [apiVersion] specific API version for request
 * @param {Object} [body] JSON-serializable body for POST requests
 * @param {string} method name of HTTP method to use
 * @param {string} path WordPress.com API path with %s and %d placeholders, e.g. /sites/%s
 * @param {Object} [query] key/value pairs for query string
 * @param {FormData} [formData] key/value pairs for POST body, encoded as "multipart/form-data"
 * @param {Object} [onSuccess] Redux action to call when request succeeds
 * @param {Object} [onFailure] Redux action to call when request fails
 * @param {Object} [onProgress] Redux action to call on progress events from an upload
 * @param {Object} [options] extra options to send to the middleware, e.g. retry policy or offline policy
 * @param {Object} [action] default action to call on HTTP events
 * @returns {Object} Redux action describing WordPress.com API HTTP request
 */
export const http = ( {
	apiVersion = 'v1',
	body,
	method,
	path,
	query = {},
	formData,
	onSuccess,
	onFailure,
	onProgress,
	...options,
}, action = null ) => ( {
	type: WPCOM_HTTP_REQUEST,
	body,
	method,
	path,
	query: method === 'GET' ? { ...query, apiVersion } : { apiVersion },
	formData,
	onSuccess: onSuccess || action,
	onFailure: onFailure || action,
	onProgress: onProgress || action,
	options,
} );
