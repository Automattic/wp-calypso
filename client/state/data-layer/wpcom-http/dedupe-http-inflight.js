/**
 * External dependencies
 */

/**
 * Internal dependencies
 */
import { processHttpRequest } from './utils';

export const requestKey = ( { path, query } ) => Object
	.keys( query || {} )
	.sort()
	.reduce( ( memo, key ) => memo + `&${ key }=${ query[ key ] }`, `path=${ path }` );

const inflightRequests = new Set();

/** We care about handling two cases of inbound http requests here
* 1. new request not inflight: add it to the inflight set and pass-through unharmed
* 2. new request already inflight: do not let the request through
*/
function handleInbound( store, next, action ) {
	const key = requestKey( action );
	if ( inflightRequests.has( key ) ) {
		return; // action should not pass go, should not collect $200
	}
	inflightRequests.add( key );
	next( action );
}

function handleOutbound( store, next, action ) {
	const key = requestKey( action );
	inflightRequests.delete( key );
	next( action );
}

export default processHttpRequest(
	handleInbound,
	handleOutbound,
);
