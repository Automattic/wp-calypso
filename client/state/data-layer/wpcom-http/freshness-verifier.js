export const CLEAN_OLDER_THAN = 3 * 60 * 1000;

/** @type {Date} holds timestamp of the last time a cleaning operation was performed */
let lastCleaningTime;

/** @type {Map} holds request action keys, and last execution times */
const history = new Map();

/**
 * Generate a deterministic key for comparing request actions
 *
 * @param   {Object} action redux action
 * @returns {String}        action key
 */
const buildActionKey = ( action ) => JSON.stringify(
	action.meta ? { ...action, meta: undefined } : action
);

/**
 * Performs a cleaning operation in the data structure used to hold action keys and execution times
 * this is useful to reduce memory usage in the structure
 * by cleaning requests that are very old and irrelevant to the freshness mechanism
 */
export const cleanHistoryIfNecessary = () => {
	const now = Date.now();
	if ( lastCleaningTime && now - lastCleaningTime < CLEAN_OLDER_THAN ) {
		return;
	}
	lastCleaningTime = now;
	history.forEach( ( lastExecutionTime, key ) => {
		if ( now - lastExecutionTime > CLEAN_OLDER_THAN ) {
			history.delete( key );
		}
	} );
};

/**
 * Abort a request if our data is fresh-enough
 *
 * @param {Function}   initiator issues actual network requests
 * @param {Number}     freshness number of milliseconds since the last execution for the request be considered fresh
 * @returns {Function}           wrapped request initiator
 */
export const initiatorWithFreshness = ( initiator, freshness ) => ( store, action, next, ignoreAction ) => {
	const lastUpdate = history.get( buildActionKey( action ) ) || -Infinity;
	const now = Date.now();

	const staleness = now - lastUpdate;

	// our data is fresher than we need it to be
	// so just skip this fetch and discard the action
	if ( staleness <= freshness ) {
		ignoreAction();
		return;
	}
	return initiator( store, action, next );
};

/**
 * Wraps onSuccess with functionality needed for freshness mechanism
 *
 * @param {Function}   onSuccess actual onSuccess function
 * @returns {Function}           wrapped onSuccess function
 */
export const onSuccessWithFreshness = ( onSuccess ) => ( store, action, next, data ) => {
	cleanHistoryIfNecessary();
	history.set( buildActionKey( action ), Date.now() );
	return onSuccess( store, action, next, data );
};
