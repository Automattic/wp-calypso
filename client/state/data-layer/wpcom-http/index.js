/** @format */

/**
 * External dependencies
 */

import { compact, get } from 'lodash';

/**
 * Internal dependencies
 */
import wpcom from 'client/lib/wp';
import { WPCOM_HTTP_REQUEST } from 'client/state/action-types';
import { extendAction } from 'client/state/utils';
import {
	processInbound as inboundProcessor,
	processOutbound as outboundProcessor,
} from './pipeline';

/**
 * Returns the appropriate fetcher in wpcom given the request method
 *
 * fetcherMap :: String -> (Params -> Query -> [Body] -> Promise)
 *
 * @param {string} method name of HTTP method for request
 * @returns {Function} the fetcher
 */
const fetcherMap = method =>
	get(
		{
			GET: wpcom.req.get.bind( wpcom.req ),
			POST: wpcom.req.post.bind( wpcom.req ),
		},
		method,
		null
	);

export const successMeta = ( data, headers ) => ( { meta: { dataLayer: { data, headers } } } );
export const failureMeta = ( error, headers ) => ( { meta: { dataLayer: { error, headers } } } );
export const progressMeta = ( { total, loaded } ) => ( {
	meta: { dataLayer: { progress: { total, loaded } } },
} );

export const queueRequest = ( processOutbound, processInbound ) => ( { dispatch }, rawAction ) => {
	const action = processOutbound( rawAction, dispatch );

	if ( null === action ) {
		return;
	}

	const { body, formData, method: rawMethod, onProgress, path, query = {} } = action;

	const method = rawMethod.toUpperCase();

	const request = fetcherMap( method )(
		...compact( [
			{ path, formData },
			{ ...query }, // wpcom mutates the query so hand it a copy
			method === 'POST' && body,
			( error, data, headers ) => {
				const {
					failures,
					nextData,
					nextError,
					nextHeaders,
					shouldAbort,
					successes,
				} = processInbound( action, { dispatch }, data, error, headers );

				if ( true === shouldAbort ) {
					return null;
				}

				return !! nextError
					? failures.forEach( handler =>
							dispatch( extendAction( handler, failureMeta( nextError, nextHeaders ) ) )
						)
					: successes.forEach( handler =>
							dispatch( extendAction( handler, successMeta( nextData, nextHeaders ) ) )
						);
			},
		] )
	);

	if ( 'POST' === method && onProgress ) {
		request.upload.onprogress = event =>
			dispatch( extendAction( onProgress, progressMeta( event ) ) );
	}
};

export default {
	[ WPCOM_HTTP_REQUEST ]: [ queueRequest( outboundProcessor, inboundProcessor ) ],
};
