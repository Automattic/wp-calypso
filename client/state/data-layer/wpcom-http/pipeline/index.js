/**
 * External dependencies
 */
import {
	compact,
	pick,
} from 'lodash';

/**
 * Internal dependencies
 */
import { applyDuplicatesHandlers, removeDuplicateGets } from './remove-duplicate-gets';

/**
 * @typedef {Object} InboundData
 * @property {Object} originalRequest the Redux action describing the inbound request
 * @property {ReduxStore} store Redux store
 * @property {*} originalData response data from returned network request
 * @property {*} originalError response error from returned network request
 * @property {*} nextData transformed response data
 * @property {*} nextError transformed response error
 * @property {Object[]} failures list of `onFailure` actions to dispatch
 * @property {Object[]} successes list of `onSuccess` actions to dispatch
 * @property {Boolean} [shouldAbort] whether or not no further processing should occur for request
 */

/**
 * @typedef {Function} InboundProcessor
 * @param {InboundData} information about request response at this stage in the pipeline
 * @returns {InboundData} contains transformed responder actions
 */

/**
 * @typedef {Object} OutboundData
 * @property {Object} originalRequest the Redux action describing the outbound request
 * @property {ReduxStore} store Redux store
 * @property {?Object} nextRequest the processed request to pass along the chain
 */

/**
 * @typedef {Function} OutboundProcessor
 * @param {OutboundData} information about request at this stage in the pipeline
 * @returns {OutboundData} contains transformed nextRequest
 */

/** @type {InboundProcessor[]} */
const inboundChain = [
	applyDuplicatesHandlers,
];

/** @type {OutboundProcessor[]} */
const outboundChain = [
	removeDuplicateGets,
];

const applyInboundProcessor = ( inboundData, nextProcessor ) =>
	inboundData.shouldAbort !== true
		? nextProcessor( inboundData )
		: inboundData;

const applyOutboundProcessor = ( outboundData, nextProcessor ) =>
	outboundData.nextRequest !== null
		? nextProcessor( outboundData )
		: outboundData;

export const processInboundChain = chain => ( originalRequest, store, originalData, originalError ) =>
	pick(
		chain.reduce( applyInboundProcessor, {
			originalRequest,
			store,
			originalData,
			originalError,
			nextData: originalData,
			nextError: originalError,
			failures: compact( [ originalRequest.onFailure ] ),
			successes: compact( [ originalRequest.onSuccess ] ),
		} ),
		[ 'failures', 'nextData', 'nextError', 'successes', 'shouldAbort' ],
	);

export const processOutboundChain = chain => ( originalRequest, store ) =>
	chain
		.reduce( applyOutboundProcessor, { originalRequest, store, nextRequest: originalRequest } )
		.nextRequest;

export const processInbound = processInboundChain( outboundChain );

export const processOutbound = processOutboundChain( inboundChain );
