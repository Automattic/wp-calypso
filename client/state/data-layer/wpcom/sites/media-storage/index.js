/**
 * Internal dependencies
 */

import { dispatchRequest } from 'calypso/state/data-layer/wpcom-http/utils';
import { http } from 'calypso/state/data-layer/wpcom-http/actions';
import {
	SITE_MEDIA_STORAGE_REQUEST,
	SITE_MEDIA_STORAGE_REQUEST_SUCCESS,
	SITE_MEDIA_STORAGE_REQUEST_FAILURE,
} from 'calypso/state/action-types';
import { receiveMediaStorage } from 'calypso/state/sites/media-storage/actions';

import { registerHandlers } from 'calypso/state/data-layer/handler-registry';

export function requestMediaStorage( action ) {
	return [
		http(
			{
				method: 'GET',
				path: `/sites/${ action.siteId }/media-storage`,
				apiVersion: '1.1',
			},
			action
		),
	];
}

export const requestMediaStorageSuccess = ( { siteId }, mediaStorage ) => [
	receiveMediaStorage( mediaStorage, siteId ),
	{ type: SITE_MEDIA_STORAGE_REQUEST_SUCCESS, siteId },
];

export const requestMediaStorageError = ( { siteId }, error ) => [
	{ type: SITE_MEDIA_STORAGE_REQUEST_FAILURE, siteId, error },
];

registerHandlers( 'state/data-layer/wpcom/sites/media-storage/index.js', {
	[ SITE_MEDIA_STORAGE_REQUEST ]: [
		dispatchRequest( {
			fetch: requestMediaStorage,
			onSuccess: requestMediaStorageSuccess,
			onError: requestMediaStorageError,
		} ),
	],
} );
