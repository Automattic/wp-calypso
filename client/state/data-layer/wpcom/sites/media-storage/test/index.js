/**
 * Internal dependencies
 */
import { requestMediaStorage, requestMediaStorageSuccess, requestMediaStorageError } from '../';
import { receiveMediaStorage } from 'calypso/state/sites/media-storage/actions';
import { http } from 'calypso/state/data-layer/wpcom-http/actions';

const ARBITRARY_SITE_ID = 12378129379;

describe( 'wpcom-api', () => {
	describe( '#requestMediaStorage()', () => {
		test( 'should return HTTP request action to media storage endpoint', () => {
			const action = { siteId: ARBITRARY_SITE_ID };

			expect( requestMediaStorage( action ) ).toEqual( [
				http(
					{ apiVersion: '1.1', method: 'GET', path: `/sites/${ ARBITRARY_SITE_ID }/media-storage` },
					action
				),
			] );
		} );
	} );

	describe( '#requestMediaStorageSuccess()', () => {
		test( 'should dispatch receive media storage actions', () => {
			const mediaStorageResponse = {
				max_storage_bytes: 214748364800,
				storage_used_bytes: 349630702,
			};
			const result = requestMediaStorageSuccess(
				{ siteId: ARBITRARY_SITE_ID },
				mediaStorageResponse
			);

			expect( result[ 0 ] ).toEqual(
				receiveMediaStorage( mediaStorageResponse, ARBITRARY_SITE_ID )
			);
			expect( result ).toMatchSnapshot();
		} );
	} );

	describe( '#requestMediaStorageError()', () => {
		test( 'should dispatch receive media storage actions', () => {
			const result = requestMediaStorageError(
				{ siteId: ARBITRARY_SITE_ID },
				new Error( 'arbitrary error' )
			);

			expect( result ).toMatchSnapshot();
		} );
	} );
} );
