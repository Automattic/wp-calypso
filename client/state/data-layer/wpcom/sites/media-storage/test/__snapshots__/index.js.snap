// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`wpcom-api #requestMediaStorageError() should dispatch receive media storage actions 1`] = `
Array [
  Object {
    "error": [Error: arbitrary error],
    "siteId": 12378129379,
    "type": "SITE_MEDIA_STORAGE_REQUEST_FAILURE",
  },
]
`;

exports[`wpcom-api #requestMediaStorageSuccess() should dispatch receive media storage actions 1`] = `
Array [
  Object {
    "mediaStorage": Object {
      "max_storage_bytes": 214748364800,
      "storage_used_bytes": 349630702,
    },
    "siteId": 12378129379,
    "type": "SITE_MEDIA_STORAGE_RECEIVE",
  },
  Object {
    "siteId": 12378129379,
    "type": "SITE_MEDIA_STORAGE_REQUEST_SUCCESS",
  },
]
`;
